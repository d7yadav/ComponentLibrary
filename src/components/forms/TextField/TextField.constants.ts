import type { TextFieldVariant, TextFieldSize, TextFieldInputType, ValidationState } from './TextField.types';

export const TEXTFIELD_VARIANTS: Record<TextFieldVariant, TextFieldVariant> = {
  filled: 'filled',
  outlined: 'outlined',
  standard: 'standard',
} as const;

export const TEXTFIELD_SIZES: Record<TextFieldSize, TextFieldSize> = {
  small: 'small',
  medium: 'medium',
} as const;

export const TEXTFIELD_INPUT_TYPES: Record<string, TextFieldInputType> = {
  text: 'text',
  email: 'email',
  password: 'password',
  number: 'number',
  tel: 'tel',
  url: 'url',
  search: 'search',
  date: 'date',
  time: 'time',
  'datetime-local': 'datetime-local',
  month: 'month',
  week: 'week',
} as const;

export const VALIDATION_STATES: Record<ValidationState, ValidationState> = {
  none: 'none',
  success: 'success',
  warning: 'warning',
  error: 'error',
} as const;

export const TEXTFIELD_SIZE_CONFIGS = {
  small: {
    height: 40,
    padding: '8px 12px',
    fontSize: '0.875rem',
    iconSize: 16,
    borderRadius: 8,
    labelPadding: '20px',
    labelTopSmall: '6px',
  },
  medium: {
    height: 48,
    padding: '12px 16px',
    fontSize: '1rem',
    iconSize: 20,
    borderRadius: 12,
    labelPadding: '25px',
    labelTopSmall: '8px',
  },
} as const;

export const VALIDATION_STATE_COLORS = {
  none: {
    border: 'divider',
    focus: 'primary.main',
    helperText: 'text.secondary',
  },
  success: {
    border: 'success.main',
    focus: 'success.main',
    helperText: 'success.main',
  },
  warning: {
    border: 'warning.main',
    focus: 'warning.main',
    helperText: 'warning.main',
  },
  error: {
    border: 'error.main',
    focus: 'error.main',
    helperText: 'error.main',
  },
} as const;

export const TEXTFIELD_ANIMATION_DURATIONS = {
  focus: '0.2s',
  blur: '0.2s',
  validation: '0.3s',
  loading: '1.5s',
} as const;

export const TEXTFIELD_ANIMATION_EASINGS = {
  spring: 'cubic-bezier(0.25, 0.46, 0.45, 0.94)',
  bounce: 'cubic-bezier(0.68, -0.55, 0.265, 1.55)',
  smooth: 'cubic-bezier(0.4, 0, 0.2, 1)',
} as const;

export const ACCESSIBILITY_CONSTANTS = {
  minTouchTarget: 44, // Minimum touch target size in pixels
  focusOutlineWidth: 2,
  focusOutlineOffset: 2,
  requiredIndicator: '*',
  loadingAriaLabel: 'Loading',
  validationDebounceDefault: 300, // milliseconds
} as const;

export const INPUT_VALIDATION_PATTERNS = {
  email: /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/,
  url: /^https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)$/,
  tel: /^[\+]?[1-9][\d]{0,15}$/,
  password: {
    minLength: 8,
    requireUppercase: true,
    requireLowercase: true,
    requireNumbers: true,
    requireSpecialChars: false,
  },
} as const;

export const CHARACTER_COUNT_CONFIGS = {
  position: 'bottom-right' as const,
  warningThreshold: 0.8, // Show warning when 80% of maxLength is reached
  errorThreshold: 1.0, // Show error when maxLength is exceeded
} as const;

export const LOADING_INDICATOR_CONFIG = {
  size: 16,
  position: 'end' as const,
  color: 'inherit' as const,
} as const;

export const DEFAULT_VALIDATION_MESSAGES = {
  required: 'This field is required',
  email: 'Please enter a valid email address',
  url: 'Please enter a valid URL',
  tel: 'Please enter a valid phone number',
  minLength: (min: number) => `Must be at least ${min} characters`,
  maxLength: (max: number) => `Must be no more than ${max} characters`,
  pattern: 'Please enter a valid value',
  number: 'Please enter a valid number',
  date: 'Please enter a valid date',
  time: 'Please enter a valid time',
} as const;

export const MULTILINE_CONFIGS = {
  defaultRows: 3,
  minRows: 1,
  maxRows: 10,
  autoResize: true,
} as const;

export const FORM_INTEGRATION_CONFIGS = {
  submitOnEnter: false,
  clearOnEscape: false,
  validateOnChange: true,
  validateOnBlur: true,
  validateOnSubmit: true,
} as const;