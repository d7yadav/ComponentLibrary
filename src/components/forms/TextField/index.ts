/**
 * @fileoverview TextField component barrel export
 * @author Dilip Yadav <dilip.sm.yadav@gmail.com>
 */

export { TextField } from './TextField';
export type { 
  TextFieldProps, 
  TextFieldVariant, 
  TextFieldSize, 
  TextFieldInputType,
  ValidationState,
  TextFieldStyleProps,
  ValidationResult,
  FormIntegrationProps
} from './TextField.types';
export {
  TEXTFIELD_VARIANTS,
  TEXTFIELD_SIZES,
  TEXTFIELD_INPUT_TYPES,
  VALIDATION_STATES,
  TEXTFIELD_SIZE_CONFIGS,
  VALIDATION_STATE_COLORS,
  TEXTFIELD_ANIMATION_DURATIONS,
  TEXTFIELD_ANIMATION_EASINGS,
  ACCESSIBILITY_CONSTANTS,
  INPUT_VALIDATION_PATTERNS,
  CHARACTER_COUNT_CONFIGS,
  DEFAULT_VALIDATION_MESSAGES,
  MULTILINE_CONFIGS,
  FORM_INTEGRATION_CONFIGS,
} from './TextField.constants';