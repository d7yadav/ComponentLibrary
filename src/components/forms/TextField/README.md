# TextField

The TextField component is a very-complex forms component that provides [brief description of functionality]. It's built with accessibility, performance, and developer experience in mind.

## Features

✅ TypeScript strict mode support
✅ Theme integration with Material-UI
✅ WCAG 2.1 AA accessibility compliance
✅ Responsive design support
✅ Comprehensive Storybook documentation
✅ Performance optimized with React.memo

## Quick Start

```typescript
import { TextField } from '@/components/forms/TextField';

const App = () => {
  return (
    <TextField>
      Hello World
    </TextField>
  );
};
```

## API Reference

No specific props documented. Extends standard HTML attributes.

## Examples

See [Examples](TextField.examples.md) for comprehensive usage examples including:

- Basic usage patterns
- Advanced configurations
- Integration with forms
- Performance optimization
- Accessibility implementation

## Accessibility

This component follows WCAG 2.1 AA guidelines and includes:

- Keyboard navigation support
- Screen reader compatibility
- High contrast mode support
- Focus management
- Proper ARIA attributes

See [AI Guide](TextField.ai-guide.md) for implementation details.

## Related Components

- [Button](../Button/README.md) - For interactive actions
- [TextField](../../forms/TextField/README.md) - For form inputs
- [Card](../../data-display/Card/README.md) - For content containers

---

*Part of the AI-Friendly Material-UI Component Library*
