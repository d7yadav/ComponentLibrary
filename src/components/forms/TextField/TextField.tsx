import { forwardRef, useState, useEffect, useCallback, memo } from 'react';
import { InputAdornment, CircularProgress } from '@mui/material';
import { 
  CheckCircle, 
  Warning, 
  Error as ErrorIcon, 
  Clear 
} from '@mui/icons-material';
import { TextFieldProps, ValidationResult } from './TextField.types';
import {
  StyledTextField,
  StyledHelperText,
  StyledLabel,
  CharacterCount,
  TextFieldContainer,
  LoadingContainer,
  ValidationIconContainer,
  RequiredIndicator,
  MultilineCharacterCount,
} from './TextField.styles';
import {
  TEXTFIELD_VARIANTS,
  TEXTFIELD_SIZES,
  VALIDATION_STATES,
  ACCESSIBILITY_CONSTANTS,
  INPUT_VALIDATION_PATTERNS,
  DEFAULT_VALIDATION_MESSAGES,
  CHARACTER_COUNT_CONFIGS,
  LOADING_INDICATOR_CONFIG,
} from './TextField.constants';

/**
 * Enhanced TextField component with comprehensive validation, accessibility, and form integration
 * 
 * Features:
 * - 3 variants: filled, outlined, standard
 * - Multiple input types with built-in validation
 * - Success, warning, error validation states
 * - Character counting and length validation
 * - Loading states and icons
 * - Accessibility compliance (WCAG 2.1 AA)
 * - React Hook Form integration
 * - Debounced validation
 * - Multiline support with auto-resize
 */
export const TextField = forwardRef<HTMLDivElement, TextFieldProps>(({
  variant = TEXTFIELD_VARIANTS.outlined,
  size = TEXTFIELD_SIZES.medium,
  type = 'text',
  value,
  defaultValue,
  placeholder,
  label,
  labelPosition = 'inside',
  helperText,
  errorText,
  successText,
  warningText,
  validationState = VALIDATION_STATES.none,
  required = false,
  disabled = false,
  readOnly = false,
  loading = false,
  fullWidth = false,
  multiline = false,
  rows,
  maxRows,
  minRows,
  startIcon,
  endIcon,
  endAction,
  maxLength,
  minLength,
  pattern,
  autoComplete,
  autoFocus = false,
  inputMode,
  showCharacterCount = false,
  validate,
  validationDebounce = ACCESSIBILITY_CONSTANTS.validationDebounceDefault,
  onChange,
  onBlur,
  onFocus,
  onKeyDown,
  onKeyUp,
  onEnterPress,
  onEscapePress,
  className,
  'aria-label': ariaLabel,
  'aria-describedby': ariaDescribedBy,
  tabIndex,
  ...other
}, ref) => {
  const [internalValue, setInternalValue] = useState(value || defaultValue || '');
  const [focused, setFocused] = useState(false);
  const [validationResult, setValidationResult] = useState<ValidationResult>({
    isValid: true,
    state: validationState,
  });
  
  // Update internal value when prop value changes
  useEffect(() => {
    if (value !== undefined) {
      setInternalValue(value);
    }
  }, [value]);
  
  // Validation function
  const validateValue = useCallback((val: string): ValidationResult => {
    // Custom validation takes precedence
    if (validate) {
      const customResult = validate(val);
      if (customResult) {
        return {
          isValid: false,
          message: customResult,
          state: 'error',
        };
      }
    }
    
    // Required validation
    if (required && !val.trim()) {
      return {
        isValid: false,
        message: DEFAULT_VALIDATION_MESSAGES.required,
        state: 'error',
      };
    }
    
    // Length validation
    if (minLength && val.length < minLength) {
      return {
        isValid: false,
        message: DEFAULT_VALIDATION_MESSAGES.minLength(minLength),
        state: 'error',
      };
    }
    
    if (maxLength && val.length > maxLength) {
      return {
        isValid: false,
        message: DEFAULT_VALIDATION_MESSAGES.maxLength(maxLength),
        state: 'error',
      };
    }
    
    // Type-specific validation
    if (val && type === 'email' && !INPUT_VALIDATION_PATTERNS.email.test(val)) {
      return {
        isValid: false,
        message: DEFAULT_VALIDATION_MESSAGES.email,
        state: 'error',
      };
    }
    
    if (val && type === 'url' && !INPUT_VALIDATION_PATTERNS.url.test(val)) {
      return {
        isValid: false,
        message: DEFAULT_VALIDATION_MESSAGES.url,
        state: 'error',
      };
    }
    
    if (val && type === 'tel' && !INPUT_VALIDATION_PATTERNS.tel.test(val)) {
      return {
        isValid: false,
        message: DEFAULT_VALIDATION_MESSAGES.tel,
        state: 'error',
      };
    }
    
    // Pattern validation
    if (val && pattern && !new RegExp(pattern).test(val)) {
      return {
        isValid: false,
        message: DEFAULT_VALIDATION_MESSAGES.pattern,
        state: 'error',
      };
    }
    
    return {
      isValid: true,
      state: 'success',
    };
  }, [validate, required, minLength, maxLength, type, pattern]);
  
  // Debounced validation
  useEffect(() => {
    if (!internalValue && !required) return;
    
    const timer = setTimeout(() => {
      const result = validateValue(String(internalValue));
      setValidationResult(result);
    }, validationDebounce);
    
    return () => clearTimeout(timer);
  }, [internalValue, validateValue, validationDebounce, required]);
  
  // Handle input change
  const handleChange = (event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    const newValue = event.target.value;
    setInternalValue(newValue);
    onChange?.(event);
  };
  
  // Handle focus
  const handleFocus = (event: React.FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    setFocused(true);
    onFocus?.(event);
  };
  
  // Handle blur
  const handleBlur = (event: React.FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    setFocused(false);
    onBlur?.(event);
  };
  
  // Handle key events
  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    if (event.key === 'Enter' && onEnterPress) {
      onEnterPress(event);
    }
    if (event.key === 'Escape' && onEscapePress) {
      onEscapePress(event);
    }
    onKeyDown?.(event);
  };
  
  // Handle clear action
  const handleClear = () => {
    setInternalValue('');
    if (onChange) {
      const event = {
        target: { value: '' },
      } as React.ChangeEvent<HTMLInputElement>;
      onChange(event);
    }
  };
  
  // Determine current validation state
  const currentValidationState = validationState !== 'none' ? validationState : validationResult.state;
  
  // Determine helper text to display
  const getHelperText = () => {
    if (errorText && currentValidationState === 'error') return errorText;
    if (warningText && currentValidationState === 'warning') return warningText;
    if (successText && currentValidationState === 'success') return successText;
    if (validationResult.message && currentValidationState === 'error') return validationResult.message;
    return helperText;
  };
  
  // Determine validation icon
  const getValidationIcon = () => {
    switch (currentValidationState) {
      case 'success':
        return <CheckCircle className="validation-icon" />;
      case 'warning':
        return <Warning className="validation-icon" />;
      case 'error':
        return <ErrorIcon className="validation-icon" />;
      default:
        return null;
    }
  };
  
  // Character count logic
  const characterCount = String(internalValue).length;
  const isCharacterWarning = maxLength ? characterCount >= maxLength * CHARACTER_COUNT_CONFIGS.warningThreshold : false;
  const isCharacterError = maxLength ? characterCount > maxLength : false;
  
  // Build start adornment
  const startAdornment = (startIcon || loading) ? (
    <InputAdornment position="start">
      {loading ? (
        <LoadingContainer position="start">
          <CircularProgress
            size={LOADING_INDICATOR_CONFIG.size}
            color={LOADING_INDICATOR_CONFIG.color}
            className="loading-spinner"
          />
        </LoadingContainer>
      ) : startIcon}
    </InputAdornment>
  ) : undefined;
  
  // Build end adornment
  const endAdornment = (endIcon || endAction || currentValidationState !== 'none' || (value && !disabled && !readOnly)) ? (
    <InputAdornment position="end">
      {endAction && !loading && endAction}
      {!endAction && endIcon && !loading && endIcon}
      {!endAction && !endIcon && currentValidationState !== 'none' && (
        <ValidationIconContainer validationState={currentValidationState}>
          {getValidationIcon()}
        </ValidationIconContainer>
      )}
      {!endAction && !endIcon && currentValidationState === 'none' && value && !disabled && !readOnly && (
        <Clear 
          onClick={handleClear}
          style={{ cursor: 'pointer', fontSize: '1rem' }}
          aria-label="Clear input"
        />
      )}
    </InputAdornment>
  ) : undefined;
  
  // Accessibility props
  const accessibilityProps = {
    'aria-label': ariaLabel,
    'aria-describedby': ariaDescribedBy,
    'aria-invalid': currentValidationState === 'error',
    'aria-required': required,
    ...(tabIndex !== undefined && { tabIndex }),
  };
  
  // Render character count
  const renderCharacterCount = () => {
    if (!showCharacterCount || !maxLength) return null;
    
    const CharacterCountComponent = multiline ? MultilineCharacterCount : CharacterCount;
    
    return (
      <CharacterCountComponent 
        component="div"
        variant="caption"
        data-testid="textfield-character-count"
        isWarning={isCharacterWarning}
        isError={isCharacterError}
      >
        {characterCount}/{maxLength}
      </CharacterCountComponent>
    );
  };
  
  const displayHelperText = getHelperText();
  
  return (
    <TextFieldContainer fullWidth={fullWidth} className={className}>
      {labelPosition === 'above' && label && (
        <StyledLabel
          component="label"
          variant="body2"
          validationState={currentValidationState}
        >
          {label}
          {required && (
            <RequiredIndicator className="required-indicator" aria-hidden="true">
              {ACCESSIBILITY_CONSTANTS.requiredIndicator}
            </RequiredIndicator>
          )}
        </StyledLabel>
      )}
      <StyledTextField
        ref={ref}
        variant={variant}
        size={size}
        type={type}
        value={internalValue}
        placeholder={placeholder}
        label={
          labelPosition === 'inside' ? (
            label && required ? (
              <>
                {label}
                <RequiredIndicator aria-hidden="true">
                  {ACCESSIBILITY_CONSTANTS.requiredIndicator}
                </RequiredIndicator>
              </>
            ) : label
          ) : undefined
        }
        disabled={disabled}
        fullWidth={fullWidth}
        multiline={multiline}
        rows={rows}
        maxRows={maxRows}
        minRows={minRows}
        autoComplete={autoComplete}
        autoFocus={autoFocus}
        inputMode={inputMode}
        InputProps={{
          startAdornment,
          endAdornment,
          readOnly,
          ...accessibilityProps,
        }}
        // Custom props for styled component
        customVariant={variant}
        customSize={size}
        validationState={currentValidationState}
        loading={loading}
        readOnly={readOnly}
        hasStartIcon={Boolean(startIcon)}
        hasEndIcon={Boolean(endIcon)}
        hasEndAction={Boolean(endAction)}
        showCharacterCount={showCharacterCount}
        focused={focused}
        labelPosition={labelPosition}
        onChange={handleChange}
        onFocus={handleFocus}
        onBlur={handleBlur}
        onKeyDown={handleKeyDown}
        onKeyUp={onKeyUp}
        {...other}
      />
      
      {displayHelperText && (
        <StyledHelperText
          component="div"
          variant="caption"
          validationState={currentValidationState}
          className={currentValidationState !== 'none' ? 'with-icon' : ''}
        >
          {displayHelperText}
        </StyledHelperText>
      )}
      
      {renderCharacterCount()}
    </TextFieldContainer>
  );
});

TextField.displayName = 'TextField';