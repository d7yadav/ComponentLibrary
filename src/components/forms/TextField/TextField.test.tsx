/**
 * @fileoverview TextField Component Tests
 * @author AI Generated Tests - 2025-07-13
 */

import { ThemeProvider , createTheme } from '@mui/material/styles';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe, toHaveNoViolations } from 'jest-axe';

import { TextField } from './TextField';
import { TEXT_FIELD_CONSTANTS } from './TextField.constants';
import type { TextFieldProps } from './TextField.types';

// Add jest-axe matcher
expect.extend(toHaveNoViolations);

// Test theme
const testTheme = createTheme();

// Test wrapper component
const TestWrapper = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={testTheme}>{children}</ThemeProvider>
);

// Helper function to render component with theme
const renderWithTheme = (ui: React.ReactElement, options = {}) => {
  return render(ui, { wrapper: TestWrapper, ...options });
};

describe('TextField Component', () => {
  // Basic rendering tests
  describe('Rendering', () => {
    it('renders without crashing', () => {
      renderWithTheme(<TextField>Test content</TextField>);
      expect(screen.getByRole('textbox')).toBeInTheDocument();
    });

    it('renders with placeholder correctly', () => {
      const placeholder = 'Enter your text here';
      renderWithTheme(<TextField placeholder={placeholder} />);
      expect(screen.getByPlaceholderText(placeholder)).toBeInTheDocument();
    });

    it('renders with custom test id', () => {
      const testId = 'custom-text-field';
      renderWithTheme(
        <TextField data-testid={testId}>Test</TextField>
      );
      expect(screen.getByTestId(testId)).toBeInTheDocument();
    });

    it('applies custom className', () => {
      const customClass = 'custom-class';
      renderWithTheme(
        <TextField className={customClass} data-testid="custom-field" />
      );
      expect(screen.getByTestId('custom-field')).toHaveClass(customClass);
    });
  });

  // Props testing
  describe('Props', () => {
    describe('variant prop', () => {
      const variants: TextFieldProps['variant'][] = [
        'filled',
        'outlined',
        'standard',
      ];

      variants.forEach((variant) => {
        it(`renders ${variant} variant correctly`, () => {
          renderWithTheme(
            <TextField variant={variant} data-testid="variant-test">
              {variant} variant
            </TextField>
          );
          const element = screen.getByTestId('variant-test');
          expect(element).toBeInTheDocument();
        });
      });
    });

    describe('size prop', () => {
      const sizes: TextFieldProps['size'][] = [
        'small',
        'medium',
      ];

      sizes.forEach((size) => {
        it(`renders ${size} size correctly`, () => {
          renderWithTheme(
            <TextField size={size} data-testid="size-test">
              {size} size
            </TextField>
          );
          const element = screen.getByTestId('size-test');
          expect(element).toBeInTheDocument();
        });
      });
    });


    it('handles loading state correctly', () => {
      renderWithTheme(
        <TextField loading label="Loading TextField" data-testid="loading-test" />
      );
      
      // Loading state should be visually indicated and field should be disabled during loading
      const element = screen.getByTestId('loading-test');
      expect(element).toBeInTheDocument();
      
      // Input should be present and labeled correctly
      const input = screen.getByRole('textbox');
      expect(input).toBeInTheDocument();
    });

    it('handles disabled state correctly', () => {
      renderWithTheme(
        <TextField disabled label="Disabled TextField" data-testid="disabled-test" />
      );
      
      // The input element should be disabled
      const input = screen.getByRole('textbox');
      expect(input).toBeDisabled();
    });
  });

  // Interaction testing
  describe('Interactions', () => {

    it('handles input changes', async () => {
      const user = userEvent.setup();
      const handleChange = vi.fn();
      
      renderWithTheme(
        <TextField onChange={handleChange} />
      );
      
      const input = screen.getByRole('textbox');
      await user.type(input, 'test input');
      expect(handleChange).toHaveBeenCalled();
    });

    it('handles focus and blur events', async () => {
      const handleFocus = vi.fn();
      const handleBlur = vi.fn();
      
      renderWithTheme(
        <TextField onFocus={handleFocus} onBlur={handleBlur} />
      );
      
      const input = screen.getByRole('textbox');
      input.focus();
      expect(handleFocus).toHaveBeenCalled();
      
      input.blur();
      expect(handleBlur).toHaveBeenCalled();
    });
  });

  // Form integration testing
  describe('Form Integration', () => {
    it('integrates with form libraries', () => {
      const handleChange = vi.fn();
      
      renderWithTheme(
        <TextField 
          name="test-field"
          value="test-value"
          onChange={handleChange}
        />
      );
      
      const element = screen.getByRole('textbox');
      expect(element).toHaveAttribute('name', 'test-field');
    });

    it('handles validation states', () => {
      renderWithTheme(
        <TextField 
          label="Validation Field"
          error
          errorText="Error message"
        />
      );
      
      expect(screen.getByText('Error message')).toBeInTheDocument();
    });
  });

  // Icon testing
  describe('Icons', () => {
    const TestIcon = () => <span data-testid="test-icon">Icon</span>;

    it('renders start icon correctly', () => {
      renderWithTheme(
        <TextField startIcon={<TestIcon />}>
          With Start Icon
        </TextField>
      );
      
      expect(screen.getByTestId('test-icon')).toBeInTheDocument();
    });

    it('renders end icon correctly', () => {
      renderWithTheme(
        <TextField endIcon={<TestIcon />}>
          With End Icon
        </TextField>
      );
      
      expect(screen.getByTestId('test-icon')).toBeInTheDocument();
    });
  });

  // Accessibility testing
  describe('Accessibility', () => {
    it('meets WCAG accessibility guidelines', async () => {
      const { container } = renderWithTheme(
        <TextField label="Accessible TextField" />
      );
      
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('supports screen readers', () => {
      const ariaLabel = 'Screen reader label';
      renderWithTheme(
        <TextField aria-label={ariaLabel}>
          TextField for screen readers
        </TextField>
      );
      
      expect(screen.getByLabelText(ariaLabel)).toBeInTheDocument();
    });

    it('supports keyboard navigation', () => {
      renderWithTheme(<TextField label="Keyboard TextField" />);
      
      const element = screen.getByRole('textbox');
      element.focus();
      expect(element).toHaveFocus();
    });

    it('provides proper ARIA attributes', () => {
      renderWithTheme(
        <TextField 
          label="ARIA TextField"
          inputProps={{
            'aria-describedby': 'description',
            'aria-controls': 'controlled-element'
          }}
        />
      );
      
      const element = screen.getByRole('textbox');
      expect(element).toHaveAttribute('aria-describedby', 'description');
      expect(element).toHaveAttribute('aria-controls', 'controlled-element');
    });
  });

  // Theme integration testing
  describe('Theme Integration', () => {
    it('respects theme customizations', () => {
      const customTheme = createTheme({
        components: {
          MuiTextField: {
            defaultProps: {
              variant: 'filled',
            },
            styleOverrides: {
              root: {
                backgroundColor: 'rgb(255, 0, 0)',
              },
            },
          },
        },
      });

      render(
        <ThemeProvider theme={customTheme}>
          <TextField data-testid="themed-text-field">
            Themed TextField
          </TextField>
        </ThemeProvider>
      );

      const element = screen.getByTestId('themed-text-field');
      expect(element).toBeInTheDocument();
    });

    it('supports dark mode', () => {
      const darkTheme = createTheme({
        palette: {
          mode: 'dark',
        },
      });

      render(
        <ThemeProvider theme={darkTheme}>
          <TextField>Dark Mode TextField</TextField>
        </ThemeProvider>
      );

      expect(screen.getByRole('textbox')).toBeInTheDocument();
    });
  });

  // Performance testing
  describe('Performance', () => {
    it('renders efficiently with many instances', () => {
      const manyComponents = Array.from({ length: 100 }, (_, i) => (
        <TextField key={i}>TextField {i}</TextField>
      ));

      const { container } = renderWithTheme(<div>{manyComponents}</div>);
      expect(container.querySelectorAll('[role="textbox"]')).toHaveLength(100);
    });

  });

  // Edge cases and error handling
  describe('Edge Cases', () => {
    it('handles empty children gracefully', () => {
      renderWithTheme(<TextField />);
      expect(screen.getByRole('textbox')).toBeInTheDocument();
    });

    it('handles null children gracefully', () => {
      renderWithTheme(<TextField>{null}</TextField>);
      expect(screen.getByRole('textbox')).toBeInTheDocument();
    });

    it('handles undefined props gracefully', () => {
      renderWithTheme(
        <TextField 
          variant={undefined as any} 
          size={undefined as any}
        >
          Undefined Props
        </TextField>
      );
      expect(screen.getByRole('textbox')).toBeInTheDocument();
    });

    it('handles complex configurations', () => {
      renderWithTheme(
        <TextField
          variant="filled"
          size="small"
          disabled={false}
          loading={false}
        >
          Complex TextField
        </TextField>
      );
      
      expect(screen.getByRole('textbox')).toBeInTheDocument();
    });
  });

  // Snapshot testing
  describe('Snapshots', () => {
    it('matches snapshot for default props', () => {
      const { container } = renderWithTheme(
        <TextField>Default TextField</TextField>
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for all variants', () => {
      const variants: TextFieldProps['variant'][] = [
        'filled',
        'outlined',
        'standard',
      ];

      variants.forEach((variant) => {
        const { container } = renderWithTheme(
          <TextField variant={variant}>{variant} TextField</TextField>
        );
        expect(container.firstChild).toMatchSnapshot(`variant-${variant}`);
      });
    });
  });
});
