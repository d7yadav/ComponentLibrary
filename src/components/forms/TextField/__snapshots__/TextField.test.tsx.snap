// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextField Component > Snapshots > matches snapshot for all variants > variant-filled 1`] = `
<div
  class="css-mx7cbl"
>
  <div
    class="MuiFormControl-root MuiTextField-root css-mz8t4m-MuiFormControl-root-MuiTextField-root"
  >
    <div
      aria-invalid="false"
      aria-required="false"
      class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-1pht7va-MuiInputBase-root-MuiFilledInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiFilledInput-input css-voecp4-MuiInputBase-input-MuiFilledInput-input"
        id="«r4u»"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`TextField Component > Snapshots > matches snapshot for all variants > variant-outlined 1`] = `
<div
  class="css-mx7cbl"
>
  <div
    class="MuiFormControl-root MuiTextField-root css-18gj8ev-MuiFormControl-root-MuiTextField-root"
  >
    <div
      aria-invalid="false"
      aria-required="false"
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
        id="«r50»"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-1nf2c5d-MuiNotchedOutlined-root"
        >
          <span
            aria-hidden="true"
            class="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`TextField Component > Snapshots > matches snapshot for all variants > variant-standard 1`] = `
<div
  class="css-mx7cbl"
>
  <div
    class="MuiFormControl-root MuiTextField-root css-19o7mrc-MuiFormControl-root-MuiTextField-root"
  >
    <div
      aria-invalid="false"
      aria-required="false"
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-5h82ro-MuiInputBase-root-MuiInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input css-1yrc8ca-MuiInputBase-input-MuiInput-input"
        id="«r52»"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`TextField Component > Snapshots > matches snapshot for default props 1`] = `
<div
  class="css-mx7cbl"
>
  <div
    class="MuiFormControl-root MuiTextField-root css-18gj8ev-MuiFormControl-root-MuiTextField-root"
  >
    <div
      aria-invalid="false"
      aria-required="false"
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
        id="«r4s»"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-1nf2c5d-MuiNotchedOutlined-root"
        >
          <span
            aria-hidden="true"
            class="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
